fem/restriction.cpp

```diff
@@ -1012,6 +1276,116 @@ void L2FaceRestriction::MultTranspose(const Vector& x, Vector& y) const
    }
 }
 
+void L2FaceRestriction::FillI(SparseMatrix &mat) const
+{
+   const int face_dofs = dof;
+   auto d_indices1 = scatter_indices1.Read();
+   auto d_indices2 = scatter_indices2.Read();
+   auto I = mat.ReadWriteI();
+   MFEM_FORALL(fdof, nf*face_dofs,
+   {
+      const int iE1 = d_indices1[fdof];
+      const int iE2 = d_indices2[fdof];
+      AddNnz(iE1,I,face_dofs);
+      AddNnz(iE2,I,face_dofs);
+   });
+}
+
+void L2FaceRestriction::FillJandData(const Vector &ea_data,
+                                     SparseMatrix &mat) const
+{
+   const int face_dofs = dof;
+   auto d_indices1 = scatter_indices1.Read();
+   auto d_indices2 = scatter_indices2.Read();
+   auto I = mat.ReadWriteI();
+   auto mat_fea = Reshape(ea_data.Read(), face_dofs, face_dofs, 2, nf);
+   auto J = mat.WriteJ();
+   auto Data = mat.WriteData();
+   MFEM_FORALL(fdof, nf*face_dofs,
+   {
+      const int f  = fdof/face_dofs;
+      const int iF = fdof%face_dofs;
+      const int iE1 = d_indices1[f*face_dofs+iF];
+      const int iE2 = d_indices2[f*face_dofs+iF];
+      const int offset1 = AddNnz(iE1,I,face_dofs);
+      const int offset2 = AddNnz(iE2,I,face_dofs);
+      //Could use threads to optimize this loop
+      for (int jF = 0; jF < face_dofs; jF++)
+      {
+         const int jE1 = d_indices1[f*face_dofs+jF];
+         const int jE2 = d_indices2[f*face_dofs+jF];
+         J[offset2+jF] = jE1;
+         J[offset1+jF] = jE2;
+         Data[offset2+jF] = mat_fea(jF,iF,0,f);
+         Data[offset1+jF] = mat_fea(jF,iF,1,f);
+      }
+   });
+}
+
+void L2FaceRestriction::FactorizeBlocks(Vector &fea_data, const int elemDofs,
+                                        const int ne, Vector &ea_data) const
+{
+   const int face_dofs = dof;
+   const int elem_dofs = elemDofs;
+   const int NE = ne;
+   if (m==L2FaceValues::DoubleValued)
+   {
+      auto d_indices1 = scatter_indices1.Read();
+      auto d_indices2 = scatter_indices2.Read();
+      auto mat_fea = Reshape(fea_data.Read(), face_dofs, face_dofs, 2, nf);
+      auto mat_ea = Reshape(ea_data.ReadWrite(), elem_dofs, elem_dofs, ne);
+      MFEM_FORALL(f, nf,
+      {
+         const int e1 = d_indices1[f*face_dofs]/elem_dofs;
+         const int e2 = d_indices2[f*face_dofs]/elem_dofs;
+         //TODO avoid recomputing iE1,iE2,jE1,jE2 by storing them?
```

Comment:
Are these still pending TODOs? If not, remove the comments.

Reply:
They are pending, they are clues for optimization.

Reply:
Okay, I can't recommend merging into master until this is resolved.

Reply:
Okay, I'll remove it all.